File: chaeyun/3_palindrome/min-del-to-make-str-unique/sol3-heap-queue.py
from collections import Counter
import heapq
# Heap / Priority Queue Approach
class Solution:
    def minDeletions(self, s: str) -> int:
        cnt = Counter(s)
        deletions = 0
        used_frequencies = set()
        
        heap = list(cnt.values())
        heapq.heapify(heap)
        
        while heap:
            freq = heapq.heappop(heap)
            while freq > 0 and freq in used_frequencies:
                freq -= 1
                deletions += 1
            used_frequencies.add(freq)
            
        return deletions

File: chaeyun/3_palindrome/min-del-to-make-str-unique/sol3-heap-queue.cpp
#include <unordered_map>
#include <unordered_set>
#include <queue>
#include <string>

class Solution {
public:
    int minDeletions(std::string s) {
        std::unordered_map<char, int> cnt;
        for (char c : s) {
            cnt[c]++;
        }
        
        int deletions = 0;
        std::unordered_set<int> used_frequencies;
        
        // Create a min heap (note the difference in C++ for this)
        std::priority_queue<int, std::vector<int>, std::greater<int>> heap;
        for (const auto& pair : cnt) {
            heap.push(pair.second);
        }
        
        while (!heap.empty()) {
            int freq = heap.top();
            heap.pop();

            while (freq > 0 && used_frequencies.find(freq) != used_frequencies.end()) {
                freq--;
                deletions++;
            }

            if (freq > 0) {
                used_frequencies.insert(freq);
            }
        }
        
        return deletions;
    }
};
